consensus/src/recovery_manager.rs

```diff
@@ -0,0 +1,151 @@
+    ) -> Result<RecoveryData> {
+        let author = proposal_msg.proposer();
+        let sync_info = proposal_msg.sync_info();
+        self.sync_up(sync_info, author).await
+    }
+
+    pub async fn process_vote_msg(&mut self, vote_msg: VoteMsg) -> Result<RecoveryData> {
+        let author = vote_msg.vote().author();
+        let sync_info = vote_msg.sync_info();
+        self.sync_up(sync_info, author).await
+    }
+
+    pub async fn sync_up(&mut self, sync_info: &SyncInfo, peer: Author) -> Result<RecoveryData> {
+        sync_info.verify(&self.epoch_state.verifier)?;
+        ensure!(
+            sync_info.highest_round() > self.last_committed_round,
+            "[RecoveryManager] Received sync info has lower round number than committed block"
+        );
+        ensure!(
+            sync_info.epoch() == self.epoch_state.epoch,
+            "[RecoveryManager] Received sync info is in different epoch than committed block"
+        );
+        let mut retriever = BlockRetriever::new(
+            self.network.clone(),
+            peer,
+            self.epoch_state
+                .verifier
+                .get_ordered_account_addresses_iter()
+                .collect(),
+        );
+        let recovery_data = BlockStore::fast_forward_sync(
+            sync_info.highest_ordered_cert(),
+            sync_info.highest_commit_cert(),
+            &mut retriever,
+            self.storage.clone(),
+            self.state_computer.clone(),
+        )
+        .await?;
+
+        Ok(recovery_data)
+    }
+
+    pub async fn start(
+        mut self,
+        mut event_rx: aptos_channel::Receiver<
+            (Author, Discriminant<VerifiedEvent>),
+            (Author, VerifiedEvent),
+        >,
+        close_rx: oneshot::Receiver<oneshot::Sender<()>>,
+    ) {
+        info!(epoch = self.epoch_state.epoch, "RecoveryManager started");
+        let mut close_rx = close_rx.into_stream();
+        loop {
+            futures::select! {
+                (peer_id, event) = event_rx.select_next_some() => {
+                    let result = match event {
+                        VerifiedEvent::ProposalMsg(proposal_msg) => {
+                            monitor!(
+                                "process_recovery",
+                                self.process_proposal_msg(*proposal_msg).await
+                            )
+                        }
+                        VerifiedEvent::VoteMsg(vote_msg) => {
+                            monitor!("process_recovery", self.process_vote_msg(*vote_msg).await)
+                        }
+                        VerifiedEvent::UnverifiedSyncInfo(sync_info) => {
+                            monitor!(
+                                "process_recovery",
+                                self.sync_up(&sync_info, peer_id).await
+                            )
+                        }
+                        unexpected_event => unreachable!("Unexpected event: {:?}", unexpected_event),
```

Comment:
I might be missing something, but what guarantees we don't receive a different message? 

especially if a node crashes and restarts in this mode, other nodes might have messages to send in response to messages from this node before the crash?

Reply:
it's not controlled by other nodes, it's an internal channel receives messages from epoch manager.

Reply:
I just had the same comment.
What if the epoch manager gets a commit message and then forward it here?